broker/broker.go

```diff
@@ -216,47 +214,36 @@ func (b *Broker) Bind(context context.Context, instanceID, bindingID string, det
 		return binding, fmt.Errorf("failed to get connection string")
 	}
 
-	dbName := makeDatabaseName(instanceID)
-	username := makeUserName(bindingID)
-	password, err := makeRandomPassword(passwordLength)
-	if err != nil {
-		return binding, err
-	}
-
-	session, err := MongoConnection(deployment.Connection.Direct[0], deployment.CACertificateBase64, b.RequireTLS)
-	if err != nil {
-		return binding, err
-	}
-	err = session.DB(dbName).UpsertUser(&mgo.User{
-		Username: username,
-		Password: password,
-		Roles:    []mgo.Role{mgo.RoleReadWrite},
-	})
-	if err != nil {
-		return binding, err
-	}
-
-	bindingURL, err := url.Parse(deployment.Connection.Direct[0])
-	if err != nil {
-		return binding, err
-	}
-	bindingURL.User = url.UserPassword(username, password)
-	bindingURL.Path = fmt.Sprintf("/%s", dbName)
-
-	// FIXME: Follow up story should fix mongo connection string handling.
-	// Right now we are hardcoding first host from the comma delimited list that Compose provides.
-	// url.Parse() parses mongo connection string wrongly and doesn't return an error
-	// so url.Port() returns port like "18899,aws-eu-west-1-portal.7.dblayer.com:18899"
-	port := strings.Split(bindingURL.Port(), ",")
+	// Different database types have different needs for setting up credentials
+	switch deploymentMeta.Type {
+	case "mongodb":
+		creds, err := newMongoCredentials(instanceID, bindingID, deployment)
+		if err != nil {
+			return binding, err
+		}
+		binding.Credentials = creds
+	case "elastic_search":
+		fallthrough
+	case "fakedb": // used by unit tests
```

Comment:
Things in this block require talking to live instances... the fakedb case allows for a kind of passthrough case (what was the default case previously). It's not ideal, but allows keeping the seperation of unit/integration tests.

In the future (when more service types are created) it would be better to create 'plugable' service interfaces ... at that stage the tests could create a FakeService thing and this case could be removed... but for now the 'fakedb' case can be used as an illutration for the generic case.